Translate 'Tutorial' page (#12)
reactjs/pl.reactjs.org: content/tutorial/tutorial.md

  - "docs/tutorial-zh-CN.html"
---

Ten samouczek nie zakłada jakiejkolwiek znajomości Reacta.

## Zanim zaczniemy się uczyć {#before-we-start-the-tutorial}

W tym samouczku skupimy się na stworzeniu niewielkiej gry. **Być może po tym zdaniu zechcesz go pominąć, bo nie zajmujesz się pisaniem gier - ale mimo wszystko daj mu szansę**. Techniki, których nauczysz się w tym poradniku są fundamentami do zbudowania *dowolnej* aplikacji w Reakcie, a zrozumienie ich zapewni Ci dogłębne poznanie tematu.

>Uwaga
>

> >Ten samouczek jest przeznaczony dla osób, które wolą **uczyć się poprzez praktykę**. Jeśli wolisz pouczyć się zagadnień od podstaw, zajrzyj do naszego [poradnika krok po kroku](/docs/hello-world.html). Wiedz jednak, że ten samouczek oraz tamten poradnik nawzajem się uzupełniają.

Author: Wolałbyś, żeby umieścić to w cudzysłowach?
`zajrzyj do rozdziału pt. "Poradnik krok po kroku"`
Czy masz może inny pomysł na tłumaczenie tego akapitu?

> >Ten samouczek jest przeznaczony dla osób, które wolą **uczyć się poprzez praktykę**. Jeśli wolisz pouczyć się zagadnień od podstaw, zajrzyj do naszego [poradnika krok po kroku](/docs/hello-world.html). Wiedz jednak, że ten samouczek oraz tamten poradnik nawzajem się uzupełniają.

Reviewer: Tak i...

```suggestion
>Niniejszy samouczek jest przeznaczony dla osób, które wolą się **uczyć poprzez praktykę**. Jeśli zależy ci na bardziej systematycznym opanowaniu teorii, zajrzyj do naszego wprowadzenia pt. ["Poradnik krok po kroku"](/docs/hello-world.html). Wiedz jednak, że niniejszy samouczek oraz poradnik nawzajem się uzupełniają.
```

Samouczek podzieliliśmy na kilka części:

* [Przygotowanie i konfiguracja](#setup-for-the-tutorial) opisuje, **jak rozpocząć**, aby być w stanie wykonywać dalsze kroki.
* [Przegląd](#overview) zawiera informacje o **podstawach** Reacta: komponentach, atrybutach i stanie.
* [Dokończenie gry](#completing-the-game) pokazuje **najczęstsze techniki** programowania w Reakcie.
* [Dodanie wehikułu czasu](#adding-time-travel) pozwala **dogłębnie zrozumieć** mocne strony Reacta.

> Nie musisz wcale przechodzić przez wszystkie części samouczka naraz, żeby wynieść z cokolwiek. Spróbuj jednak dojść najdalej jak możesz - nawet jeśli będą to tylko dwa rozdziały.

Author: Nawet podany przez Ciebie link sugeruje, że jednak powinno być `naraz` :-)

> W trakcie wykonywania poleceń możesz kopiować kod do edytora, ale zalecamy pisać go samodzielnie. Pozwoli Ci to wyrobić pamięć mięśniową i lepiej zrozumieć temat.

Author: To może coś takiego?
```suggestion
W trakcie wykonywania poleceń możesz kopiować kod do edytora, ale zalecamy pisanie go samodzielnie. Pozwoli Ci to bardziej skupić się na poszczególnych zagadnieniach, a także lepiej je przyswoić i dogłębniej zrozumieć.
```

> W trakcie wykonywania poleceń możesz kopiować kod do edytora, ale zalecamy pisać go samodzielnie. Pozwoli Ci to wyrobić pamięć mięśniową i lepiej zrozumieć temat.

Reviewer: W mojej popszedniej propozycji usiłowałem wysłowić, to że zazwyczaj warunkiem niezbędnym "przyswojenia" i "zrozumienia" danego zagadnienia jest właśnie skoncentrowanie na nim uwagi, "skupienie się".

> W trakcie wykonywania poleceń możesz kopiować kod do edytora, ale zalecamy pisać go samodzielnie. Pozwoli Ci to wyrobić pamięć mięśniową i lepiej zrozumieć temat.

Author: Ok, to może coś takiego?
```suggestion
W trakcie wykonywania poleceń możesz kopiować kod do edytora, ale zalecamy pisanie go samodzielnie. Pozwoli ci to bardziej skupić się na poszczególnych zagadnieniach, dzięki czemu lepiej je przyswoisz i dogłębniej zrozumiesz.
```

> W trakcie wykonywania poleceń możesz kopiować kod do edytora, ale zalecamy pisać go samodzielnie. Pozwoli Ci to wyrobić pamięć mięśniową i lepiej zrozumieć temat.

Reviewer: We're almost there :-)
Trudne w wymowie zestawienie "dogłębniej zrozumiesz". Przyznaję, to może być kwestia subiektywnych preferencji.

```suggestion
W trakcie wykonywania poleceń możesz kopiować kod do edytora, ale zalecamy pisanie go samodzielnie. Pozwoli ci to bardziej skupić się na poszczególnych zagadnieniach, dzięki czemu bardziej je sobie przyswoisz, lepiej zrozumiesz.
```

### Co będziemy budować? {#what-are-we-building}

W tym samouczku pokażemy, jak przy pomocy Reacta zbudować interaktywną grę w "kółko i krzyżyk".

Na koniec nasz kod będzie wyglądał tak: **[Efekt końcowy](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**. Kod wygląda bez sensu albo nie rozumiesz jego składni? Bez obaw! Celem tego samouczka jest właśnie pomóc Ci zrozumieć Reacta i jego składnie.

> Przed dalszą lekturą radzimy uruchomić grę w "kółko i krzyżyk". Jedną z funkcjonalności, znajdującą się na prawo od planszy, jest numerowana lista. Służy ona do wyświetlania historii wszystkich ruchów, jakie wykonano podczas rozgrywki, i jest aktualizowana na bieżąco.

Author: Wydaje mi się, że tak, bo "jakie wykonano podczas rozgrywki" to wtrącenie, które otacza się przecinkami po obydwóch stronach. Ale może @marcinostrouch będzie umiał tutaj pomóc.

> Przed dalszą lekturą radzimy uruchomić grę w "kółko i krzyżyk". Jedną z funkcjonalności, znajdującą się na prawo od planszy, jest numerowana lista. Służy ona do wyświetlania historii wszystkich ruchów, jakie wykonano podczas rozgrywki, i jest aktualizowana na bieżąco.

Reviewer: Może:

```suggestion
Zanim przystąpisz do dalszej lektury, zachęcamy cię, abyś spróbował uruchomić grę w "kółko i krzyżyk". Zwróć uwagę, że jedną z funkcjonalności tej aplikacji jest, znajdująca się na prawo od planszy, numerowana lista. Wyświetla ona na bieżąco wszystkie wykonywane w danej rozgrywce ruchy.
```

Gdy już zapoznasz się z grą, możesz ją wyłączyć. Naukę rozpoczniemy od prostego szablonu. Naszym następnym krokiem będzie konfiguracja środowiska, aby nadawało się do stworzenia gry.

### Wymagania {#prerequisites}

> Zakładamy, że HTML i JavaScript nie są Ci obce, ale nawet jeśli to Twoja pierwsza styczność z tymi językami, to samouczek nie powinien przysporzyć Ci większego problemu. Zakładamy również, że znasz takie pojęcia z programowania obiektowego jak "funkcja", "obiekt" czy "tablica", a także przynajmniej w niewielkim stopniu rozumiesz, czym jest "klasa".

Reviewer 2: Ładnie użyte cudzysłowy! Popieram! 👍

Jeśli potrzebujesz odświeżyć sobie wiedzę o JavaScripcie, zalecamy lekturę [tego poradnika](https://developer.mozilla.org/pl/docs/Web/JavaScript/Ponowne_wprowadzenie_do_JavaScript). Zwróć uwagę, że będziemy tu korzystać z niektórych funkcjonalności standardu ES6 - jednej z ostatnich wersji języka JavaScript - takich jak [funkcje strzałkowe](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Functions/Funkcje_strzalkowe) (ang. *arrow functions*), [klasy](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Classes) oraz wyrażenia [`let`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let) i [`const`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const). Możesz skorzystać ze środowiska [Babel REPL](babel://es5-syntax-example), aby sprawdzić, co powstanie w wyniku transpilowania kodu napisanego w standardzie ES6.

## Konfiguracja pod samouczek {#setup-for-the-tutorial}

> Istnieją dwa sposoby na ukończenie tego samouczka: możesz pisać kod w przeglądarce lub możesz zestawić lokalne środowisko na swoim komputerze.

Author: Masz rację, trochę mnie poniosło ;-) Chciałem, żeby to zdanie brzmiało bardziej dewelopersko.

### Opcja konfiguracyjna nr 1: Pisanie kodu w przeglądarce {#setup-option-1-write-code-in-the-browser}

To zdecydowanie najszybszy i najłatwiejszy sposób na rozpoczęcie przygody z Reactem!

Na początek otwórz w nowej karcie **[szablon startowy](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)**. Twoim oczom powinna ukazać się pusta plansza do gry w "kółko i krzyżyk" oraz panel z kodem reactowym. W tym samouczku będziemy edytować kod w tymże panelu.

Możesz teraz pominąć opcję nr 2 i przejść do rozdziału pt. ["Przegląd"](#overview), w którym przedstawiliśmy ogólne informacje o Reakcie.

### Opcja konfiguracyjna nr 2: Lokalne środowisko programistyczne {#setup-option-2-local-development-environment}

Ten sposób jest całkowicie opcjonalny i nie jest konieczny do ukończenia tego samouczka!

<br>

<details>

<summary><b>Opcjonalne: Instrukcje na temat pisania kodu w wybranym przez siebie edytorze tekstu</b></summary>

> Ta konfiguracja wymaga więcej wysiłku, lecz pozwala na ukończenie samouczka przy użyciu wybranego przez siebie edytora. Należy kolejno:

Author: Żeby pozbyć się bezokoliczników, a jednocześnie brzmieć nadal "swojsko", proponuję:
```suggestion
Ta konfiguracja wymaga więcej wysiłku, lecz pozwala na ukończenie samouczka przy użyciu dowolnego edytora. W tym celu:
```
I kolejne kroki napisane w formie "Sprawdź...", "Wykonaj...".

1. Sprawdzić, czy posiada się najnowszą wersję [Node.js](https://nodejs.org/en/).
2. Wykonać [instrukcje instalacji "Create React App"](/docs/create-a-new-react-app.html#create-react-app) w celu zainicjalizowania nowego projektu.

```bash
npx create-react-app moja-aplikacja
```

3. Usunąć wszystkie pliki z folderu `src/` w nowo utworzonym projekcie

> Uwaga:
>
>**Nie usuwaj całego folderu `src`, tylko znajdujące się wewnątrz pliki.** W następnym kroku podmienimy domyślne pliki źródłowe przykładami z samouczka.

```bash
cd moja-aplikacja
cd src

# Jeśli używasz Maca lub Linuxa, wpisz:
rm -f *

# Lub jeśli używasz Windowsa:
del *

# A następnie cofnij się do folderu z projektem
cd ..
```

4. W folderze `src/` dodaj plik `index.css` i wstaw do niego [ten kod CSS](https://codepen.io/gaearon/pen/oWWQNa?editors=0100).

5. W folderze `src/` dodaj plik `index.js` i wstaw do niego [ten kod JS](https://codepen.io/gaearon/pen/oWWQNa?editors=0010).

> 6. Wstaw te trzy linie na początku pliku `index.js` znajdującego się w folderze `src/`:

Author: Raczej spotyka się naprzemiennie: "zmień coś w linii 43" i "zmień coś w wierszu 43". Wydaje mi się, że "wiersz" brzmi tu bardziej na miejscu.

  }
}

// Przykład użycia: <ShoppingList name="Marek" />
```

Do tych śmiesznych znaczników XML-owych wrócimy za chwilę. Komponenty pozwalają wytłumaczyć Reactowi, co chcemy ujrzeć na ekranie. Gdy zmienią się nasze dane, React w sposób efektywny zaktualizuje i ponownie wyrenderuje komponenty.

W powyższym przykładzie `ShoppingList` jest **reactowym komponentem klasowym**. Na wejściu komponent przyjmuje parametry, nazywane "atrybutami" (ang. *props*, skrót od *properties*), i przy pomocy metody `render` zwraca strukturę widoków do wyświetlenia.

> Metoda `render` zwraca *opis* tego, co zostanie wyświetlone na ekranie. React bierze ten opis i wyświetla jego wynik końcowy. Precyzyjniej rzecz ujmując, metoda `render` zwraca **element reactowy**, który jest lekką reprezentacją tego, co zostanie wyświetlone na ekranie. Większość programistów reactowych używa specjalnej składni zwanej "JSX", która ułatwia pisanie wspomnianych struktur. Składnia `<div />` jest przekształcana podczas budowania w kod: `React.createElement('div')`. Zatem powyższy przykład jest równoznaczny z:

Author: Akurat tutaj wygenerowany obiekt (a właściwie struktura obiektów) może być całkiem skomplikowana. Chodzi o to, że reprezentacja jest zwięzła, minimalna.

## Analizowanie szablonu startowego {#inspecting-the-starter-code}

Jeśli zamierzasz przejść przez ten samouczek **w przeglądarce**, otwórz ten link w nowej karcie: **[Szablon startowy](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)**. Jeśli zamierzasz pracować **lokalnie**, zamiast tego otwórz plik `src/index.js`, znajdujący się w folderze z projektem (jego tworzenie opisywaliśmy w rozdziale pt. ["Konfiguracja pod samouczek"](#setup-option-2-local-development-environment)).

Wspomniany "szablon startowy" będzie bazą dla naszej aplikacji. Dodaliśmy już do niego style CSS, więc teraz wystarczy, że skupisz się na nauce Reacta i implementowaniu gry w "kółko i krzyżyk".

Analizując kod, łatwo zauważyć, że mamy do czynienia z trzema komponentami reactowymi:

* Square (pole)
* Board (plansza)
* Game (gra)

> Komponent `Square` (pole) renderuje pojedynczy element `<button>`, a `Board` (plansza) renderuje 9 takich pól. Z kolei komponent `Game` (gra) renderuje planszę wypełnioną symbolami zastępczymi, które zmodyfikujemy w późniejszym etapie. Aplikacja w obecnym stanie nie posiada żadnej interaktywności.

Reviewer 3: Chyba nie ma sensu na siłę zmieniać tego zdania, `interaktywne komponenty` są bardziej self-explanatory, niż `interaktywność`

```suggestion
Komponent `Square` (pole) renderuje pojedynczy element `<button>`, a `Board` (plansza) renderuje 9 takich pól. Z kolei komponent `Game` (gra) renderuje planszę wypełnioną symbolami zastępczymi, które zmodyfikujemy w późniejszym etapie. Aplikacja w obecnym stanie nie posiada żadnych interaktywnych komponentów.
```

}
```

Przed zmianą:

![React Devtools](../images/tutorial/tictac-empty.png)

Po zmianie: W każdym wyrenderowanym polu powinien być wyświetlany kolejny numer.

![React Devtools](../images/tutorial/tictac-numbers.png)

**[Zobacz dotychczasowy kod](https://codepen.io/gaearon/pen/aWWQOG?editors=0010)**

Gratulacje! Właśnie udało Ci się "przekazać atrybut" z komponentu nadrzędnego `Board` do komponentu potomnego `Square`. Przekazywanie atrybutów to jeden ze sposobów na przepływ danych między rodzicem a dzieckiem w drzewie.

### Tworzenie interaktywnego komponentu {#making-an-interactive-component}

Niech teraz komponent `Square` wypełnia się literą "X" po kliknięciu w niego.
Najpierw dokonaj zmian w elemencie `<button />` zwracanym przez `Square` w metodzie `render()`:

```javascript{4}
class Square extends React.Component {
  render() {
    return (

>       <button className="square" onClick={function() { alert('klik'); }}>

Reviewer 3: ```suggestion
      <button className="square" onClick={function() { alert('kliknięcie'); }}>
```

Author: W "alercie" raczej umieściłbym "klik" niż "kliknięcie". Ewentualnie "kliknięto w przycisk".

Reviewer 2: "kliknięto w przycisk" 👍

        {this.props.value}
      </button>
    );
  }
}
```

Teraz gdy klikniemy na polu, przeglądarka powinna wyświetlić wiadomość w oknie dialogowym.

>Uwaga
>
>Żeby oszczędzić sobie pisania i zapobiec [mylącemu zachowaniu `this`](https://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/), od teraz obsługę zdarzeń będziemy zapisywać za pomocą [funkcji strzałkowych](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Functions/Funkcje_strzalkowe) (ang. *arrow functions*):
>
>```javascript{4}
>class Square extends React.Component {
>  render() {
>    return (
>      <button className="square" onClick={() => alert('klik')}>
>        {this.props.value}
>      </button>
>    );
>  }
>}
>```
>

> >Zauważ, że za pomocą `onClick={() => alert('klik')}` pod atrybutem `onClick` przekazujemy *funkcję*. Zostanie ona wywołana dopiero po kliknięciu w przycisk. Częstym błędem jest zapominanie o `() =>` i pisanie `onClick={alert('klik')}`, co powoduje wyświetlenie wiadomości w momencie renderowania komponentu.

Reviewer 3: ```suggestion
>Zauważ, że za pomocą `onClick={() => alert('kliknięcie')}` pod atrybutem `onClick` przekazujemy *funkcję*. Zostanie ona wywołana dopiero po kliknięciu w przycisk. Częstym błędem jest zapominanie o `() =>` i pisanie `onClick={alert('klik')}`, co powoduje wyświetlenie wiadomości w momencie renderowania komponentu.
```

Author: To samo, co powyżej.

Reviewer 2: "kliknięto w przycisk" 👍

> W następnym kroku sprawimy, by komponent `Square` "pamiętał", że został kliknięty, i wyświetlał literę "X". Do "pamiętania" rzeczy komponenty używają **stanu**.

Reviewer 3: Za pierwszym przeczytaniem zastanawiałem się, co to są `rzeczy komponenty` i miałem wrażenie, że to literówka (`rzeczy komponentu`), więc zmiana struktury wyeliminuje to ryzyko.

```suggestion
W następnym kroku sprawimy, by komponent `Square` "pamiętał", że został kliknięty, i wyświetlał literę "X". Komponenty do "pamiętania" rzeczy używają **stanu**.
```

  render() {
    return (

>       <button className="square" onClick={() => alert('klik')}>

Reviewer 3: ```suggestion
      <button className="square" onClick={() => alert('kliknięcie')}>
```

        {this.props.value}
      </button>
    );
  }
}
```

>Uwaga
>

> >W [klasach javascriptowych](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Classes) należy zawsze wywoływać specjalną metodę `super` w konstruktorze klasy potomnej. Dlatego wszystkie komponenty reactowe będące klasą, która mają zdefiniowany konstruktor powinny w nim na początku wywoływać `super(props)`.

Reviewer 3: ```suggestion
>W [klasach javascriptowych](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Classes) należy zawsze wywoływać specjalną metodę `super` w konstruktorze klasy potomnej. Dlatego wszystkie komponenty reactowe będące klasą, które mają zdefiniowany konstruktor powinny w nim na początku wywoływać `super(props)`.
```

> >W [klasach javascriptowych](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Classes) należy zawsze wywoływać specjalną metodę `super` w konstruktorze klasy potomnej. Dlatego wszystkie komponenty reactowe będące klasą, które mają zdefiniowany konstruktor, powinny w nim na początku wywoływać `super(props)`.

Reviewer 2: ```suggestion
>W [klasach javascriptowych](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Classes) należy zawsze wywoływać metodę `super` w konstruktorze klasy potomnej. Dlatego wszystkie komponenty reactowe będące klasą, które mają zdefiniowany konstruktor, powinny w nim na początku wywoływać `super(props)`.
```

Teraz zmienimy kod w metodzie `render` komponentu `Square` tak, aby wyświetlał po kliknięciu wartość aktualnego stanu:

> * Zamień `this.props.value` na `this.state.value` wewnątrz znacznika `<button>`.

Reviewer 3: W *Lifting state up* używaliśmy przy `<input>` słowa element, więc może trzymajmy to spójnie?

```suggestion
* Zamień `this.props.value` na `this.state.value` wewnątrz elementu `<button>`.
```

Author: Element to coś, co trafia do drzewa DOM i jest na stronie. Znacznik to sposób zapisu (kod). Tutaj podmieniamy coś w kodzie, więc zostawiłbym "znacznika".

* Zamień uchwyt zdarzenia `() => alert()` na `() => this.setState({value: 'X'})`.
* Umieść atrybuty `className` i `onClick` w osobnych liniach dla poprawy czytelności kodu.

> Po powyższych zmianach znacznik `<button>`, zwracany przez komponent `Square`, powinien wyglądać następująco:

Reviewer 3: ```suggestion
Po powyższych zmianach element `<button>`, zwracany przez komponent `Square`, powinien wyglądać następująco:
```

Author: Tutaj już bardziej pasuje "element".

}
```

> Wywołując `this.setState` z wnętrza uchwytu dla zdarzenia `onClick` w metodzie `render`, mówimy Reactowi, aby ponownie renderował element `<button>` po każdym kliknięciu. Po aktualizacji, wartość zmiennej `this.state.value` będzie równa `"X"`, dlatego też `X` zostanie wyświetlone na planszy. Kliknięcie na dowolne pole powinno spowodować wyświetlenie w nim litery "X".

Reviewer 2: ```suggestion
Wywołując `this.setState` z wnętrza uchwytu dla zdarzenia `onClick` w metodzie `render`, mówimy Reactowi, aby po każdym kliknięciu, ponownie renderował element `<button>`. Po aktualizacji, wartość zmiennej `this.state.value` będzie równa `"X"`, dlatego też `X` zostanie wyświetlone na planszy. Kliknięcie na dowolne pole powinno spowodować wyświetlenie w nim litery "X".
```

Gdy w komponencie wywołujesz `setState`, React automatycznie aktualizuje również wszystkie komponenty znajdujące się poniżej w hierarchii.

**[Zobacz dotychczasowy kod](https://codepen.io/gaearon/pen/VbbVLg?editors=0010)**

### Narzędzia deweloperskie {#developer-tools}

Rozszerzenie o nazwie "React Devtools" dla przeglądarek [Chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=pl) i [Firefox](https://addons.mozilla.org/pl/firefox/addon/react-devtools/) pozwala na zbadanie drzewa komponentów reactowych za pomocą narzędzi deweloperskich wbudowanych w przeglądarkę.

<img src="../images/tutorial/devtools.png" alt="React Devtools" style="max-width: 100%">

Dzięki temu rozszerzeniu możesz podejrzeć atrybuty i stan dowolnego komponentu w drzewie.

> Po zainstalowaniu rozszerzenia wystarczy, że klikniesz prawym przyciskiem myszy na dowolnym elemencie strony i wybierzesz z menu "Zbadaj". W pasku zakładek, jako ostatnia po prawej, powinna pojawić się zakładka "React".

Author: Akurat u mnie (w wersji polskiej) mam "Zbadaj". Dodam jeszcze do tego wersję angielską, żeby każdy miał coś dla siebie :-)
![image](https://user-images.githubusercontent.com/30927218/53274161-94558c80-36f6-11e9-8b79-4b338980a1fc.png)

### Wyciąganie stanu w górę {#lifting-state-up}

W obecnej wersji aplikacji każdy z komponentów `Square` decyduje o własnym stanie. Jednak do rozstrzygnięcia gry potrzebny nam będzie jeden wspólny stan, przechowujący wartości dla wszystkich 9 pól na planszy.

Być może przeszło Ci przez myśl, że to komponent `Board` powinien "pytać" każdy z komponentów `Square` o jego stan. Nawet jeśli jest to możliwe w Reakcie, odradzamy takiego podejścia, ponieważ kod stanie się wtedy bardziej zagmatwany, podatny na błędy i trudny w utrzymaniu. Zamiast tego należy przenieść stan całej planszy do komponentu nadrzędnego - `Board`. Komponent ten, poprzez atrybuty, będzie mówił każdemu z komponentów `Square`, co mają wyrenderować - podobnie jak w jednym z poprzednich podrozdziałów, gdzie [przekazywaliśmy kolejne numery do komponentów `Square`](#passing-data-through-props).

**Aby móc zebrać dane z wielu komponentów potomnych lub umożliwić dwóm potomkom komunikowanie się ze sobą, należy zadeklarować ich wspólny stan w rodzicu. Taki rodzic może wtedy przekazać poszczególne wartości potomkom poprzez atrybuty; dzięki temu potomkowie będą zsynchronizowani zarówno ze sobą nawzajem, jak i z rodzicem.**

> Wynoszenie stanu w górę struktury to dość częsty zabieg podczas refaktoryzacji (ang. *refactoring*) kodu. Wykorzystajmy zatem okazję do wypróbowania tego schematu. Dodajmy konstruktor do komponentu `Board` i ustawmy w nim stan przechowujący tablicę 9 wartości `null`. Każdy z elementów tej tablicy będzie odpowiadał jednemu polu na planszy:

Reviewer: ```suggestion
Wynoszenie stanu w górę struktury to dość częsty zabieg podczas refaktoryzacji (ang. *refactoring*) kodu. Wykorzystajmy zatem okazję do wypróbowania tego schematu. Dodajmy konstruktor do komponentu `Board` i ustawmy w nim stan przechowujący tablicę dziewięciu wartości `null`. Każdy z elementów tej tablicy będzie odpowiadał jednemu polu na planszy:
```

]
```

Aktualnie metoda `renderSquare` komponentu `Board` wygląda następująco:

```javascript
  renderSquare(i) {
    return <Square value={i} />;
  }
```

Na początku tego samouczka [przekazaliśmy atrybut `value`](#passing-data-through-props) z komponentu `Board`, aby wyświetlić liczby od 0 do 8 w kolejnych polach planszy. W innym rozdziale zastąpiliśmy liczby literą "X", [określaną na podstawie własnego stanu komponentu `Square`](#making-an-interactive-component). To właśnie dlatego `Square` obecnie ignoruje atrybut `value` przekazywany z nadrzędnego komponentu `Board`.

> Teraz z powrotem wrócimy do mechanizmu przekazywania atrybutów. Zmienimy kod komponentu `Board` tak, aby mówił poszczególnym komponentom `Square`, jaką mają wartość (`"X"`, `"O"` lub `null`). Zdefiniowaliśmy już tablicę `squares` (pol. *pola*) w konstruktorze komponentu `Board`, więc teraz zostaje nam zmodyfikować metodę `renderSquare` (pol. *renderuj pole*) tak, aby odczytywała z niej wartości:

Reviewer: "z powrotem wrócimy"?

```suggestion
Teraz ponownie zastosujemy mechanizm przekazywania atrybutów. Zmienimy kod komponentu `Board` tak, aby mówił poszczególnym komponentom `Square`, jaką mają wartość (`"X"`, `"O"` lub `null`). Zdefiniowaliśmy już tablicę `squares` (pol. *pola*) w konstruktorze komponentu `Board`, więc teraz zostaje nam zmodyfikować metodę `renderSquare` (pol. *renderuj pole*) tak, aby odczytywała z niej wartości:
```

```javascript{2}
  renderSquare(i) {
    return <Square value={this.state.squares[i]} />;
  }
```

**[Zobacz dotychczasowy kod](https://codepen.io/gaearon/pen/gWWQPY?editors=0010)**

> Każde pole otrzyma poprzez atrybut `value` odpowiednio: `"X"`, `"O"` lub `null` (jeśli pole jest puste).

Reviewer: ```suggestion
Każde pole otrzyma poprzez atrybut `value` odpowiednio: `"X"`, `"O"` lub `null` (w przypadku pola pustego).
```

> Następnie musimy zmienić zachowanie komponentu `Square` po kliknięciu na nim, ponieważ teraz to `Board` decyduje, które pola są wypełnione. Musimy zatem określić sposób, w jaki komponent `Square` może aktualizować stan swojego rodzica. Jako że stan jest dla komponentu `Board` czymś prywatnym, nie możemy tak po prostu nadpisać jego wartości z poziomu potomka `Square`.

Reviewer: ```suggestion
Następnie musimy zmienić zachowanie komponentu `Square` po kliknięciu na nim, ponieważ teraz to `Board` decyduje, które pola są wypełnione. Musimy zatem określić sposób, w jaki komponent `Square` może aktualizować stan swojego rodzica. Jako że stan komponentu `Board` należy tylko do niego, nie możemy tak po prostu nadpisać jego wartości z poziomu potomka `Square`.
```

> Aby zapewnić prywatność stanowi komponentu `Board`, przekażemy do `Square` odpowiednią funkcję poprzez atrybuty. Funkcja ta będzie wywoływana za każdym razem, gdy ktoś kliknie na pole. Zmieńmy więc metodę `renderSquare` w następujący sposób:

Reviewer: ```suggestion
Aby zapewnić komponentowi `Board` wyłączność na własność jego stanu, do komponentu `Square`  przekażemy odpowiednią funkcję za pomocą atrybutów. Funkcja ta będzie wywoływana za każdym razem, gdy ktoś kliknie na pole. Zmieńmy więc metodę `renderSquare` w następujący sposób:
```

Author: Faktycznie, "zapewnienie prywatności komponentowi" dosyć mocno go personifikuje. Nie podglądajmy komponentu, jak śpi ;-)

  }
```

>Uwaga
>
>Podzieliliśmy kod zwracanego elementu na kilka linii, aby poprawić czytelność. Dodaliśmy również nawiasy po to, aby JavaScript niepotrzebnie nie wstawił średnika za instrukcją `return`, powodując tym samym błąd.

Przekazujemy teraz z komponentu `Board` do `Square` dwa atrybuty: `value` i `onClick`. Atrybut `onClick` jest funkcją, którą `Square` może wywołać po kliknięciu. Żeby tak się stało, należy:

* zamienić `this.state.value` na `this.props.value` w metodzie `render` komponentu `Square`,
* zamienić `this.setState()` na `this.props.onClick()` w metodzie `render` komponentu `Square`,
* usunąć konstruktor z klasy `Square`, ponieważ nie będzie ona już przechowywała stanu gry.

> Po powyższych zmianach kod komponentu `Square` powinien wyglądać następująco:

Reviewer: jw.

```suggestion
Po wprowadzeniu powyższych zmian kod komponentu `Square` powinien wyglądać następująco:
```

}
```

Gdy użytkownik kliknie na pole, zostanie wywołana funkcja `onClick` dostarczona przez komponent `Board`. Dzieje się to dlatego, że:

> 1. Atrybut `onClick` wbudowanego komponentu DOM `<button>` informuje Reacta, żeby ustawił obserwator (ang. *listener*) na kliknięcie.

Reviewer: Jeśli "obserwator kliknięć" nie jest przyjętym powszechnie tłumaczeniem "event listener" to propnuję oddać to bardziej opisowo lub pójść w stronę "słuchacza".

```suggestion
1. Atrybut `onClick` wbudowanego komponentu DOM `<button>` informuje Reacta, zaczął nasłuchiwać kliknięć użytkownika.
```

2. Gdy użytkownik kliknie na przycisk, React wywoła uchwyt zdarzenia `onClick` zdefiniowany w metodzie `render()` komponentu `Square`.
3. Uchwyt ten wywoła funkcję `this.props.onClick()`, czyli atrybut przekazany przez komponent `Board`.
4. Ponieważ komponent `Board` przekazał swojemu potomkowi atrybut `onClick={() => this.handleClick(i)}`, kliknięcie w `Square` spowoduje w konsekwencji wywołanie `this.handleClick(i)` wewnątrz komponentu `Board`.
5. Nie zdefiniowaliśmy jeszcze metody `handleClick()`, dlatego aplikacja w tym momencie przestanie działać.

>Uwaga
>
>Atrybut `onClick` elementu DOM `<button>` ma dla Reacta specjalne znaczenie, ponieważ jest to wbudowany komponent. W przypadku własny komponentów, jak `Square`, nazwa tego atrybutu może być dowolna. Moglibyśmy nazwać inaczej zarówno atrybut `onClick` w `Square`, jak i `handleClick` w `Board`. Przyjęło się jednak określać atrybuty odpowiedzialne za wywołanie zdarzenia jako `on[Event]`, a uchwyty obsługujące zdarzenia jako `handle[Event]`.

> Gdy spróbujemy kliknąć na pole planszy, naszym oczom ukaże się błąd, ponieważ nie zdefiniowaliśmy jeszcze metody `handleClick`. Dodajmy ją zatem:

Reviewer: Podobało mi się "oczom ukaże się" użyte po raz pierwszy wcześniej w tłumaczeniu jako jednorazowy, spontaniczny wykwit barwnego języka. Proponuję jedanak nie powtarzać tego wyrażenia, a dać tutaj coć bardziej potocznego:

```suggestion
Jeśli teraz klikniemy na pole planszy, otrzymamy błąd, ponieważ nie zdefiniowaliśmy jeszcze metody `handleClick`. Dodajmy ją zatem:
```

}
```

**[Zobacz dotychczasowy kod](https://codepen.io/gaearon/pen/ybbQJX?editors=0010)**

> Po zastosowaniu powyższych zmian znów będziemy mogli klikać w pola planszy. Różnica polega na tym, że teraz stan planszy jest przechowywany w komponencie `Board` zamiast być w każdym z komponentów `Square` z osobna. Gdy stan w `Board` ulegnie zmianie, wszystkie pola zostaną ponownie wyrenderowane. Przechowywanie stanu wszystkich pól w jednym miejscu pozwoli nam w przyszłości wyłonić zwycięzcę rozgrywki.

Reviewer: ```suggestion
Po zastosowaniu powyższych zmian znów będziemy mogli klikać w pola planszy. Różnica polega na tym, że teraz stan planszy jest przechowywany w komponencie `Board` nie zaś w każdym z komponentów `Square` z osobna. Gdy stan w `Board` ulegnie zmianie, wszystkie pola zostaną ponownie wyrenderowane. Przechowywanie stanu wszystkich pól w jednym miejscu pozwoli nam w przyszłości wyłonić zwycięzcę rozgrywki.
```

Przenieśliśmy stan z komponentu `Square` do `Board`, przez co każdy `Square` otrzymuje od rodzica swoją wartość i informuje go o kliknięciu. W terminologii reactowej komponenty `Square` są teraz **komponentami kontrolowanymi**. `Board` ma nad nimi pełną kontrolę.

> Zauważ, że w metodzie `handleClick` używamy `.slice()` do stworzenia kopii tablicy `squares`, zamiast ją modyfikować bezpośrednio. Odpowiedź dlaczego znajdziesz w następnym podrozdziale.

Reviewer: ```suggestion
Zauważ, że w metodzie `handleClick` używamy `.slice()` do stworzenia kopii tablicy `squares`, zamiast ją modyfikować bezpośrednio. W następnym podrozdziale dowiesz się dlaczego.
```

### Dlaczego niezmienność jest istotna {#why-immutability-is-important}

> W poprzednim przykładzie zasugerowaliśmy użycie metody `.slice()` do stworzenia kopii tablicy `squares`, zamiast ją bezpośrednio modyfikować. Omówimy teraz niezmienność (ang. *immutability*) oraz powód, dla którego warto ją znać.

Reviewer: ```suggestion
W poprzednim przykładzie zasugerowaliśmy użycie metody `.slice()` do stworzenia kopii tablicy `squares`, zamiast ją bezpośrednio modyfikować. Omówimy teraz bardzo ważne zagadnienie niezmienności (ang. *immutability*).
```

          <Board />
        </div>
        <div className="game-info">
          <div>{/* Status */}</div>
          <ol>{/* Do zrobienia */}</ol>
        </div>
      </div>
    );
  }
}
```

> Następnie przekażemy komponentowi `Board` atrybuty `squares` oraz `onClick` od jego rodzica. Ponieważ będziemy odtąd przekazywać tylko jeden obserwator zdarzenia "onClick" obsługujący wiele pól, musimy go wywoływać z parametrem określającym lokalizację klikniętego pola. Można to zrobić w trzech krokach:

Reviewer: ```suggestion
Następnie przekażemy komponentowi `Board` atrybuty `squares` oraz `onClick` od jego rodzica. Ponieważ będziemy odtąd przekazywać tylko jednego obserwatora zdarzenia "onClick" obsługujący wiele pól, musimy go wywoływać z parametrem określającym lokalizację klikniętego pola. Można to zrobić w trzech krokach:
```

Author: Wydaje mi się, że analogicznie do słowa "klient" (nieos.), który odmienia się "klienty" (a nie: "klienci"), można tutaj zastosować "jeden obserwator" (nieos.) zamiast "jednego obserwatora".

Reviewer: O rety, naprawdę "klienty"? Nie będę się spierał jeśli faktycznie tak się przyjęło tłumaczyć w literaturze branżowj, ale...

Wydaje mi się, że forma "klienty" oraz "...przekazywać obserwator" stoją w bardzo dużym kontraście do gramatycznych przyzwyczajeń użytkownika języka polskiego. Można powiedzieć, że formy te atakują owe przyzwyczajenia swoją gramatyczną odmiennością. Mamy przecież odpowiednie formy liczby mnogiej dla tych rzeczowników.

Author: Rada Językowa zabrała stanowisko w tej sprawie: http://www.rjp.pan.pl/index.php?option=com_content&view=article&id=676:klienci-agenci-czy-klienty-agenty&catid=44&Itemid=58

W tym momencie komponent `Board` będzie potrzebował już tylko dwóch metod: `renderSquare` oraz `render`. Stan gry oraz metoda `handleClick` powinny znaleźć się w komponencie `Game`.

**[Zobacz dotychczasowy kod](https://codepen.io/gaearon/pen/EmmOqJ?editors=0010)**

### Wyświetlanie ruchów z przeszłości {#showing-the-past-moves}

Dzięki temu, że zapisujemy historię stanów gry w "kółko i krzyżyk", możemy wyświetlić na ekranie ruchy z przeszłości w formie listy.

W poprzednich rozdziałach dowiedzieliśmy się, że elementy reactowe są [pierwszoklasowymi obiektami javascriptowymi](https://pl.wikipedia.org/wiki/Typ_pierwszoklasowy) (ang. *first-class JavaScript objects*); możemy je przekazywać w aplikacji do woli. Do wyrenderowania listy w Reakcie możemy wykorzystać tablicę elementów reactowych.

Javascriptowe tablice mają wbudowaną [metodę `map()`](https://developer.mozilla.org/pl/docs/Web/JavaScript/Referencje/Obiekty/Array/map), której często używa się do przekształcania danych na inne, na przykład:

```js
const numbers = [1, 2, 3];
const doubled = numbers.map(x => x * 2); // [2, 4, 6]
``` 

> Przy pomocy metody `map` możemy przekształcić historię ruchów w listę elementów reactowych reprezentujących przyciski na ekranie, a następnie wyświetlić tę listę na ekranie, by umożliwić "przeskakiwanie" między ruchami.

Reviewer: ```suggestion
Przy pomocy metody `map` możemy przekształcić historię ruchów w listę elementów reactowych reprezentujących przyciski na ekranie, a następnie wyświetlić tę listę, aby umożliwić "przeskakiwanie" między ruchami.
```

  }
```

**[Zobacz dotychczasowy kod](https://codepen.io/gaearon/pen/EmmGEa?editors=0010)**

> Dla każdego z ruchów w historii gry tworzymy element listy `<li>` zawierający przycisk `<button>`. Do przycisku przekazujemy obserwator zdarzenia `onClick`, który wywołuje metodę `this.jumpTo()` (której jeszcze nie zaimplementowaliśmy). Powinniśmy być teraz w stanie zobaczyć listę wykonanych już ruchów oraz ostrzeżenie w konsoli narzędzi deweloperskich o treści:

Reviewer: ```suggestion
Dla każdego z ruchów w historii gry tworzymy element listy `<li>` zawierający przycisk `<button>`. Do przycisku przekazujemy obserwatora zdarzenia `onClick`, który wywołuje metodę `this.jumpTo()` (której jeszcze nie zaimplementowaliśmy). Powinniśmy być teraz w stanie zobaczyć listę wykonanych już ruchów oraz ostrzeżenie w konsoli narzędzi deweloperskich o treści:
```

Author: Jak we wcześniejszym komentarzu, zostawiłbym "obserwator" (nieos.).

Reviewer: OK

>  Warning:
>  Each child in an array or iterator should have a unique "key" prop. Check the render method of "Game".
>
>  (Ostrzeżenie:
>  Każdy potomek w tablicy lub iteratorze powinien posiadać unikalną wartość atrybutu "key". Sprawdź metodę render komponentu "Game".)

Ustalmy, co oznacza powyższe ostrzeżenie.

### Wybieranie odpowiedniego klucza {#picking-a-key}

> Gdy renderujemy listę, React zapamiętuje pewne informacje o każdym z wyrenderowanych jej elementów. Kiedy zaktualizujemy listę, React musi wiedzieć, jak określić, które z elementów uległy zmianie. Moglibyśmy przecież dodać, usunąć, przesunąć lub zmienić zawartość dowolnego z elementów listy.

Reviewer: ```suggestion
Gdy renderujemy listę, React zapamiętuje pewne informacje o każdym z jej wyrenderowanych elementów. Kiedy zaktualizujemy listę, React musi wiedzieć, jak określić, które z elementów uległy zmianie. Moglibyśmy przecież dodać, usunąć, przesunąć lub zmienić zawartość dowolnego z elementów listy.
```

Wyobraź sobie, że taka lista:

```html
<li>Alicja: pozostało 7 zadań</li>
<li>Bartek: pozostało 5 zadań</li>
```

zmieniłaby się w taką:

```html
<li>Bartek: pozostało 9 zadań</li>
<li>Czarek: pozostało 8 zadań</li>
<li>Alicja: pozostało 5 zadań</li>
```

> Człowiek byłby z łatwością w stanie zauważyć, że oprócz zmiany wartości liczbowych, zamieniliśmy miejscami Alicję z Bartkiem, a także "wcisnęliśmy" Czarka pomiędzy nich. Niestety React jest tylko programem komputerowym i nie zna naszych intencji. Z tego powodu musimy jawnie określić atrybut *key* dla każdego z elementów listy, który wyróżnia go spośród sąsiednich elementów. Jedną z możliwych wartości mógłby być ciąg znaków, np. `alicja`, `bartek`, `czarek`. Gdyby te dane pochodziły z bazy danych, każda z powyższych osób miałaby zapewne przyporządkowany unikalny identyfikator, którego można by użyć jako klucza.

Reviewer: ```suggestion
Człowiek czytający tę listę może z łatwością zauważyć, że oprócz zmiany wartości liczbowych, zamieniliśmy miejscami Alicję z Bartkiem, a także "wcisnęliśmy" Czarka pomiędzy nich. Niestety React jest tylko programem komputerowym i nie zna naszych intencji. Z tego powodu musimy jawnie określić atrybut *key* dla każdego z elementów listy, który wyróżnia go spośród sąsiednich elementów. Jedną z możliwych wartości mógłby być ciąg znaków, np. `alicja`, `bartek`, `czarek`. Gdyby te dane pochodziły z bazy danych, każda z powyższych osób miałaby zapewne przyporządkowany unikalny identyfikator, którego można by użyć jako klucza.
```

```html
<li key={user.id}>{user.name}: pozostało {user.taskCount} zadań</li>
```

> Gdy lista jest ponownie renderowana, React odczytuje wartość klucza każdego z elementów i wyszukuje po niej odpowiadające elementy na poprzedniej liście. Jeśli nowa lista zawiera klucz, którego wcześniej nie było, element jest tworzony. Jeśli na nowej liście nie ma klucza, który był poprzednio, element jest usuwany. Jeśli uda się dopasować klucze na obydwóch listach, odpowiadający im element jest odpowiednio przesuwany. Klucze informują Reacta o tożsamości każdego z elementów, co pozwala na zarządzanie stanem pomiędzy kolejnymi renderowaniami. Jeśli klucz elementu się zmieni, React usunie ten element i stworzy go na nowo - z nową wartością klucza i nowym stanem.

Reviewer: którego wcześniej **na niej** było, element jest tworzony. Jeśli na nowej liście nie ma klucza, który był **na niej** poprzednio,

```suggestion
Gdy lista jest ponownie renderowana, React odczytuje wartość klucza każdego z elementów i wyszukuje po niej odpowiadające elementy na poprzedniej liście. Jeśli nowa lista zawiera klucz, którego wcześniej na niej było, element jest tworzony. Jeśli na nowej liście nie ma klucza, który był na niej poprzednio, element jest usuwany. Jeśli uda się dopasować klucze na obydwóch listach, odpowiadający im element jest odpowiednio przesuwany. Klucze informują Reacta o tożsamości każdego z elementów, co pozwala na zarządzanie stanem pomiędzy kolejnymi renderowaniami. Jeśli klucz elementu się zmieni, React usunie ten element i stworzy go na nowo - z nową wartością klucza i nowym stanem.
```

`key` jest w Reakcie specjalną i zarezerwowaną właściwością (tak jak `ref` - bardziej zaawansowana funkcjonalność). W momencie gdy element jest tworzony po raz pierwszy, React odczytuje wartość jego atrybutu `key` i zachowuje ją bezpośrednio w renderowanym elemencie. Mimo iż `key` wygląda jak jeden z atrybutów, nie można go odczytać z `this.props.key`. React używa właściwości `key` automatycznie, aby zdecydować, które komponenty wymagają aktualizacji. Zaś sam komponent nie jest w stanie odczytać wartości własnego `key`.

**Zalecamy przypisywanie odpowiednich kluczy do każdej tworzonej listy dynamicznej**. Jeśli nie masz do dyspozycji wartości, która nadawałaby się na klucz, zastanów się, czy nie możesz lepiej ustrukturyzować danych.

> Jeśli nie podasz klucza, React wyświetli ostrzeżenie w konsoli i domyślnie użyje indeksu elementu. Używanie indeksu jako klucza może przysporzyć Ci kłopotów, gdy zechcesz zmienić kolejność lub dodać/usunąć elementy do listy. Jawne przekazanie `key={i}` "uciszy" ostrzeżenie, jednak problem nadal będzie występował. Takie podejście nie jest zalecane w większości przypadków.

Reviewer: ```suggestion
Jeśli nie podasz klucza, React wyświetli ostrzeżenie w konsoli i domyślnie użyje indeksu elementu. Używanie indeksu jako klucza może przysporzyć ci kłopotów, gdy zechcesz zmienić kolejność lub dodać/usunąć elementy do listy. Jawne przekazanie `key={i}` "uciszy" ostrzeżenie, jednak problem nadal będzie występował. Takie podejście nie jest zalecane w większości przypadków.
```
